### ISSUE TITLE
Add Bencher to track benchmarks in CI

### ISSUE THREAD

**Issue by epompeii** [epage/_rust#25]
@epage thank you for [the mention](https://news.ycombinator.com/item?id=42079137) earlier this week. I wanted to follow up as a lot of progress has been made since we last spoke. Bencher can now create all dimensions on-the-fly, so there is no more need to make them in advance. Further, you can now configure everything, including [setting thresholds](https://bencher.dev/docs/explanation/thresholds/) using the CLI. This should make it possible to use the same template across repos. The only thing to parameterize is the project slug.

Are you still using Divan as your benchmarking harness? @nvzqz [mentioned back in June](https://github.com/bencherdev/bencher/issues/321#issuecomment-2198512736) that the JSON output should be coming "soon".

Bencher now also supports [tracking compile times](https://bencher.dev/docs/how-to/track-build-time/) and [tracking binary size](https://bencher.dev/docs/how-to/track-file-size/), if either of those are also of interest.

**Comment by epage** [comment]
Glad to hear!

Looking over the steps, it sounds like there is still another per-repository step:

> Make sure you have [created an API token](https://bencher.dev/docs/tutorial/quick-start/#create-an-api-token) and set it as a Repository secret named BENCHER_API_TOKEN before continuing on! Navigate to Your Repo -> Settings -> Secrets and variables -> Actions -> New repository secret. Name the secret BENCHER_API_TOKEN and set the secret value to your API token.

So I'll still need to deploy a secret to all of my repos or (more ideally) manage a secret per repository, right?

**Comment by epompeii** [comment]
> So I'll still need to deploy a secret to all of my repos or (more ideally) manage a secret per repository, right?

Yes, good catch. You will need to set the `BENCHER_API_TOKEN`. Currently, the API tokens are tied to the user, so the same token should work across all projects that you create. There are plans to add project scoped API tokens in the future: https://github.com/bencherdev/bencher/issues/262

> Do you have `hyperfine` support?

Yep! Check out [the `shell_hyperfine` adapter docs](https://bencher.dev/docs/explanation/adapters/#_%EF%B8%8F-shell-hyperfine).

> I wonder if I could switch projects like https://github.com/rosetta-rs/argparse-rosetta-rs over to this?

You definitely could. However, I will note that bake-off/head-to-head comparisons isn't a use case that I have spent much time optimizing Bencher for. I'll have to add your repo to [my benchmark comparison prior art list](https://bencher.dev/docs/reference/prior-art/#benchmark-comparisons).

**Comment by epage** [comment]
Thanks for the help!

While not a fan of manually managing all of this (having an "app" take care of a lot of this is a big help), I'm giving this a try.

As I work off of a single template, I tried to generalize as much as I could.  I'm seeing a `--project` flag but no `--organization`.  I was hoping I could use `"${{ github.repository }}"` for this but it fails
- https://github.com/clap-rs/clap/blob/master/.github/workflows/bench-baseline.yml#L39
- https://github.com/clap-rs/clap/actions/runs/11805441897/job/32887903228

**Comment by epompeii** [comment]
Ah, yeah `"${{ github.repository }}"` fails because the value that is expected there is [the project slug or UUID](https://bencher.dev/docs/explanation/bencher-run/#--project-project).

My thought was to [define a configuration variable](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#defining-configuration-variables-for-multiple-workflows) for each project.

So your template could stay the same: `"${{ vars.BENCHER_PROJECT }}"`

And you just plug in the right slug for each project on the backend. Similar to how `BENCHER_API_TOKEN` is handled, but no need for the project slug to be a secret.

**Comment by epage** [comment]
For a given project slug, how does it know what organization that project is in?

Also, per repo settings are a borderline dealbreaker. I'm trying this out for now but I don't think I want to set 30 tokens to get this to work, let alone other variables.

**Comment by epompeii** [comment]
> For a given project slug, how does it know what organization that project is in?

Bencher projects are not namespaced. That is their slugs are globally unique.
Note that there can be multiple projects with the same display name, but they will have different slugs.

> Also, per repo settings are a borderline dealbreaker. I'm trying this out for now but I don't think I want to set 30 tokens to get this to work, let alone other variables

I appreciate the candid feedback on this point. So far, most users have only had a handful of repos they're integrating with. I can definitely see how the current setup doesn't scale well across dozens of repos.

**Comment by epage** [comment]
>      Bencher projects are not namespaced. That is their slugs are globally unique.
Note that there can be multiple projects with the same display name, but they will have different slugs.

So what is an "organization"?  And what if I have ambiguous repo names across github organizations?

**Comment by epompeii** [comment]
> So what is an \"organization\"?

An Organization is really just a collection of Projects, similar to GitHub. Currently, the permissions for collaborators are only available at the Organization level. As a user, you can create as many organizations as you like.

Unlike GitHub, all Bencher Organizations are private. You have to [have `view` permissions](https://bencher.dev/docs/api/organizations/organizations/#get-v0organizationsorganization) to see any details.

> And what if I have ambiguous repo names across github organizations?

Do you mean what if you have two repos named `foo`, one under namespace `one` and the second under namespace `two`?

What you do here is up to you.

A couple of options:

1. Name them `one/foo` and `two/foo`
    with the slugs `one-foo` and `two-foo`.
2. Name the both `foo`
    with the slugs `one-foo` and `two-foo`.

If you create a project named `foo`, and a Project with the slug `foo` already exists, then, Bencher just slaps a random suffix on the end to make it unique, `foo-abcd1234`. You can always change the slug for a Project (or any resource for that matter). If you use [the API directly](https://bencher.dev/docs/api/organizations/projects/#post-v0organizationsorganizationprojects) you can include a preferred slug in the call to try and avoid such a collision.
